File: app/schemas/fund.py
"""
基金相关Schema
"""
from datetime import datetime, date
from typing import Optional, List
from decimal import Decimal
from pydantic import BaseModel, Field


class FundSearchRequest(BaseModel):
    """基金搜索请求"""
    keyword: str = Field(..., description="搜索关键词", min_length=1)
    page: int = Field(default=1, ge=1, description="页码")
    size: int = Field(default=20, ge=1, le=100, description="每页数量")
    type: Optional[str] = Field(default=None, description="基金类型过滤")


class FundListItem(BaseModel):
    """基金列表项"""
    fund_code: str = Field(..., description="基金代码")
    fund_name: str = Field(..., description="基金名称")
    fund_type: str = Field(..., description="基金类型")
    purchase_status: str = Field(..., description="申购状态")
    current_scale: Optional[Decimal] = Field(default=None, description="当前规模")
    
    class Config:
        from_attributes = True


class LatestNav(BaseModel):
    """最新净值"""
    nav_date: date = Field(..., description="净值日期")
    unit_nav: Decimal = Field(..., description="单位净值")
    accumulated_nav: Decimal = Field(..., description="累计净值")
    daily_growth_rate: Optional[Decimal] = Field(default=None, description="日增长率")


class FundDetail(BaseModel):
    """基金详情"""
    fund_code: str = Field(..., description="基金代码")
    fund_name: str = Field(..., description="基金名称")
    fund_type: str = Field(..., description="基金类型")
    fund_manager: Optional[str] = Field(default=None, description="基金经理")
    establish_date: date = Field(..., description="成立日期")
    current_scale: Optional[Decimal] = Field(default=None, description="当前规模")
    management_fee: Optional[Decimal] = Field(default=None, description="管理费率")
    custody_fee: Optional[Decimal] = Field(default=None, description="托管费率")
    purchase_status: str = Field(..., description="申购状态")
    redemption_status: str = Field(..., description="赎回状态")
    benchmark: Optional[str] = Field(default=None, description="业绩基准")
    is_qdii: bool = Field(default=False, description="是否QDII基金")
    share_class: Optional[str] = Field(default=None, description="份额类别")
    latest_nav: Optional[LatestNav] = Field(default=None, description="最新净值")
    
    class Config:
        from_attributes = True


class NavData(BaseModel):
    """净值数据"""
    nav_date: date = Field(..., description="净值日期")
    unit_nav: Decimal = Field(..., description="单位净值")
    accumulated_nav: Decimal = Field(..., description="累计净值")
    daily_growth_rate: Optional[Decimal] = Field(default=None, description="日增长率")


class NavHistoryResponse(BaseModel):
    """净值历史响应"""
    fund_code: str = Field(..., description="基金代码")
    fund_name: str = Field(..., description="基金名称")
    nav_data: List[NavData] = Field(..., description="净值数据列表")


class HoldingItem(BaseModel):
    """持仓项"""
    stock_code: Optional[str] = Field(default=None, description="股票代码")
    stock_name: Optional[str] = Field(default=None, description="股票名称")
    holding_ratio: Optional[Decimal] = Field(default=None, description="持仓占比")
    holding_value: Optional[Decimal] = Field(default=None, description="持仓市值")
    industry: Optional[str] = Field(default=None, description="所属行业")


class HoldingsResponse(BaseModel):
    """持仓响应"""
    fund_code: str = Field(..., description="基金代码")
    report_date: date = Field(..., description="报告期")
    holdings: List[HoldingItem] = Field(..., description="持仓列表")
    industry_distribution: dict = Field(..., description="行业分布")


class FeeItem(BaseModel):
    """费率项"""
    holding_period: str = Field(..., description="持有期")
    purchase_fee: str = Field(..., description="申购费率")
    redemption_fee: str = Field(..., description="赎回费率")
    total_fee: str = Field(..., description="总费率")
    breakeven: str = Field(..., description="盈亏平衡点")


class FeesResponse(BaseModel):
    """费率响应"""
    fund_code: str = Field(..., description="基金代码")
    fees: List[FeeItem] = Field(..., description="费率列表")

File: frontend/app/services/fund.ts
import { get, type ApiResponse, type PageResponse } from './api'

/**
 * 基金基本信息
 */
export interface FundInfo {
  fundCode: string
  fundName: string
  fundType: string
  fundManager: string
  fundCompany: string
  establishDate: string
  fundScale: number
  benchmark: string
  riskLevel: string
  status: string
}

/**
 * 基金搜索结果项
 */
export interface FundSearchItem {
  fundCode: string
  fundName: string
  fundType: string
  fundManager: string
  fundCompany: string
  fundScale: number
  yield1Year: number
  yield3Year: number
  yield5Year: number
}

/**
 * 基金详情
 */
export interface FundDetail extends FundInfo {
  currentNav: number
  previousNav: number
  navDate: string
  yield1Day: number
  yield1Week: number
  yield1Month: number
  yield3Month: number
  yield6Month: number
  yield1Year: number
  yield3Year: number
  yield5Year: number
  yieldFromInception: number
  maxDrawdown: number
  sharpeRatio: number
  volatility: number
  description: string
  investmentStrategy: string
}

/**
 * 净值历史记录
 */
export interface NavHistory {
  date: string
  nav: number
  accumulatedNav: number
  dailyYield: number
}

/**
 * 基金持仓信息
 */
export interface FundHoldings {
  stockPositions: StockPosition[]
  bondPositions: BondPosition[]
  industryDistribution: IndustryDistribution[]
  topHoldings: TopHolding[]
  reportDate: string
}

/**
 * 股票持仓
 */
export interface StockPosition {
  stockCode: string
  stockName: string
  shares: number
  value: number
  proportion: number
}

/**
 * 债券持仓
 */
export interface BondPosition {
  bondCode: string
  bondName: string
  value: number
  proportion: number
}

/**
 * 行业分布
 */
export interface IndustryDistribution {
  industry: string
  proportion: number
}

/**
 * 十大重仓
 */
export interface TopHolding {
  rank: number
  name: string
  code: string
  proportion: number
}

/**
 * 基金费率信息
 */
export interface FundFees {
  purchaseFee: FeeRule[]
  redemptionFee: FeeRule[]
  managementFee: number
  custodyFee: number
  serviceFee: number
}

/**
 * 费率规则
 */
export interface FeeRule {
  amountRange: string
  feeRate: number
  minFee: number
  maxFee: number
}

/**
 * 搜索基金请求参数
 */
export interface SearchFundsParams {
  keyword: string
  page?: number
  size?: number
  fundType?: string
  sortBy?: string
  sortOrder?: 'asc' | 'desc'
}

/**
 * 搜索基金
 * @param keyword - 搜索关键词（基金代码或名称）
 * @param page - 页码，默认为 1
 * @param size - 每页数量，默认为 10
 * @returns 分页搜索结果
 */
export async function searchFunds(
  keyword: string,
  page: number = 1,
  size: number = 10
): Promise<ApiResponse<PageResponse<FundSearchItem>>> {
  const params: Record<string, string | number> = {
    keyword,
    page,
    size,
  }
  return get<PageResponse<FundSearchItem>>('/funds/search', { params })
}

/**
 * 获取基金详情
 * @param fundCode - 基金代码
 * @returns 基金详细信息
 */
export async function getFundDetail(
  fundCode: string
): Promise<ApiResponse<FundDetail>> {
  return get<FundDetail>(`/funds/${fundCode}`)
}

/**
 * 获取基金净值历史
 * @param fundCode - 基金代码
 * @param startDate - 开始日期 (YYYY-MM-DD)
 * @param endDate - 结束日期 (YYYY-MM-DD)
 * @returns 净值历史记录列表
 */
export async function getFundNavHistory(
  fundCode: string,
  startDate?: string,
  endDate?: string
): Promise<ApiResponse<NavHistory[]>> {
  const params: Record<string, string> = {}
  if (startDate) {
    params.startDate = startDate
  }
  if (endDate) {
    params.endDate = endDate
  }
  return get<NavHistory[]>(`/funds/${fundCode}/nav-history`, { params })
}

/**
 * 获取基金持仓信息
 * @param fundCode - 基金代码
 * @returns 持仓详细信息
 */
export async function getFundHoldings(
  fundCode: string
): Promise<ApiResponse<FundHoldings>> {
  return get<FundHoldings>(`/funds/${fundCode}/holdings`)
}

/**
 * 获取基金费率信息
 * @param fundCode - 基金代码
 * @returns 费率详细信息
 */
export async function getFundFees(
  fundCode: string
): Promise<ApiResponse<FundFees>> {
  return get<FundFees>(`/funds/${fundCode}/fees`)
}

/**
 * 获取基金排行榜
 * @param period - 排行周期 (1m/3m/6m/1y/3y/5y)
 * @param fundType - 基金类型
 * @param limit - 返回数量
 * @returns 基金排行列表
 */
export async function getFundRanking(
  period: string = '1y',
  fundType?: string,
  limit: number = 20
): Promise<ApiResponse<FundSearchItem[]>> {
  const params: Record<string, string | number> = {
    period,
    limit,
  }
  if (fundType) {
    params.fundType = fundType
  }
  return get<FundSearchItem[]>('/funds/ranking', { params })
}

/**
 * 获取基金对比数据
 * @param fundCodes - 基金代码列表
 * @returns 对比数据
 */
export async function compareFunds(
  fundCodes: string[]
): Promise<ApiResponse<FundDetail[]>> {
  return get<FundDetail[]>('/funds/compare', {
    params: { codes: fundCodes.join(',') },
  })
}

export const fundService = {
  searchFunds,
  getFundDetail,
  getFundNavHistory,
  getFundHoldings,
  getFundFees,
  getFundRanking,
  compareFunds,
}

export default fundService
